engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/IrPersistance.java

```diff
@@ -0,0 +1,760 @@
+      out.writeInline(List.class, obj.exports());
+      out.writeInline(List.class, obj.bindings());
+      out.writeBoolean(obj.isPrivate());
+      out.writeInline(Option.class, obj.location());
+      out.writeInline(MetadataStorage.class, obj.passData());
+      out.writeInline(DiagnosticStorage.class, obj.diagnostics());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected Module readObject(Input in) throws IOException, ClassNotFoundException {
+      List<Import> imports = in.readInline(List.class);
+      List<Export> exports = in.readInline(List.class);
+      List<Definition> bindings = in.readInline(List.class);
+      var isPrivate = in.readBoolean();
+      Option<IdentifiedLocation> location = in.readInline(Option.class);
+      var meta = in.readInline(MetadataStorage.class);
+      var diag = in.readInline(DiagnosticStorage.class);
+      return new Module(imports, exports, bindings, isPrivate, location, meta, diag);
+    }
+  }
+
+  @ServiceProvider(service = Persistance.class)
+  public static final class PersistMetadataStorage extends Persistance<MetadataStorage> {
+    public PersistMetadataStorage() {
+      super(MetadataStorage.class, false, 301);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected void writeObject(MetadataStorage obj, Output out) throws IOException {
+      var map =
+          obj.map(
+              (processingPass, metadata) -> {
+                var t = new Tuple2<>(processingPass.getClass().getName(), metadata);
+                return t;
+              });
+      out.writeInline(scala.collection.immutable.Map.class, map);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected MetadataStorage readObject(Input in) throws IOException, ClassNotFoundException {
+      var storage = new MetadataStorage(nil());
+      var map = in.readInline(scala.collection.immutable.Map.class);
+      var it = map.iterator();
+      while (it.hasNext()) {
+        var obj = (Tuple2<String, ProcessingPass.Metadata>) it.next();
+        try {
+          var pass = (ProcessingPass) Class.forName(obj._1()).getField("MODULE$").get(null);
+          var data = obj._2();
+          if (in.lookup(CompilerStub.class) instanceof CompilerStub stub) {
+            var restored = data.restoreFromSerialization(stub);
+            if (restored.isEmpty()) {
+              throw new IOException("Cannot convert " + data + " during deserialization");
+            } else {
+              data = restored.get();
+            }
+          } else {
+            throw new IOException("No CompilerContext registered for deserialization");
+          }
+          storage.update(pass, data);
+        } catch (ReflectiveOperationException ex) {
+          throw new IOException(ex);
+        }
+      }
+      return storage;
+    }
+  }
+
+  @ServiceProvider(service = Persistance.class)
+  public static final class PersistDiagnosticStorage extends Persistance<DiagnosticStorage> {
```

Comment:
If we know that this will always deser to Nil, then why declare it in the first place?

Reply:
Because it is referenced at various `case class` constructors and the annotation processor for `@Persistable` generates read/write code for the class.